core/trino-main/src/test/java/io/trino/operator/TestStreamingExchangeClientBuffer.java

```diff
@@ -0,0 +1,225 @@
+            assertPageEquals(buffer.pollPage(), pages.get(2));
+            assertFalse(buffer.isFinished());
+            assertFalse(buffer.isBlocked().isDone());
+            assertEquals(buffer.getRetainedSizeInBytes(), 0);
+            assertEquals(buffer.getMaxRetainedSizeInBytes(), pages.get(1).getRetainedSizeInBytes() + pages.get(2).getRetainedSizeInBytes());
+            assertEquals(buffer.getRemainingBufferCapacityInBytes(), DataSize.of(1, KILOBYTE).toBytes());
+
+            buffer.taskFinished(tasks.get(1));
+            assertTrue(buffer.isFinished());
+            assertTrue(buffer.isBlocked().isDone());
+        }
+    }
+
+    @Test
+    public void testClose()
+    {
+        StreamingExchangeClientBuffer buffer = new StreamingExchangeClientBuffer(directExecutor(), DataSize.of(1, KILOBYTE));
+        buffer.addTask(tasks.get(0));
+        buffer.addTask(tasks.get(1));
+
+        assertFalse(buffer.isFinished());
+        assertFalse(buffer.isBlocked().isDone());
+        assertNull(buffer.pollPage());
+
+        buffer.close();
+
+        assertTrue(buffer.isFinished());
+        assertTrue(buffer.isBlocked().isDone());
+        assertNull(buffer.pollPage());
+    }
+
+    @Test
+    public void testIsFinished()
+    {
+        try (StreamingExchangeClientBuffer buffer = new StreamingExchangeClientBuffer(directExecutor(), DataSize.of(1, KILOBYTE))) {
+            assertFalse(buffer.isFinished());
+            assertFalse(buffer.isBlocked().isDone());
+
+            buffer.noMoreTasks();
+
+            assertTrue(buffer.isFinished());
+            assertTrue(buffer.isBlocked().isDone());
+        }
+
+        try (StreamingExchangeClientBuffer buffer = new StreamingExchangeClientBuffer(directExecutor(), DataSize.of(1, KILOBYTE))) {
+            assertFalse(buffer.isFinished());
+            assertFalse(buffer.isBlocked().isDone());
+
+            buffer.addTask(tasks.get(0));
+            buffer.noMoreTasks();
+
+            assertFalse(buffer.isFinished());
+            assertFalse(buffer.isBlocked().isDone());
+
+            buffer.taskFinished(tasks.get(0));
+
+            assertTrue(buffer.isFinished());
+            assertTrue(buffer.isBlocked().isDone());
+        }
+
+        try (StreamingExchangeClientBuffer buffer = new StreamingExchangeClientBuffer(directExecutor(), DataSize.of(1, KILOBYTE))) {
+            assertFalse(buffer.isFinished());
+            assertFalse(buffer.isBlocked().isDone());
+
+            buffer.addTask(tasks.get(0));
+
+            assertFalse(buffer.isFinished());
+            assertFalse(buffer.isBlocked().isDone());
+
+            RuntimeException error = new RuntimeException();
+            buffer.taskFailed(tasks.get(0), error);
+
+            assertTrue(buffer.isFinished());
+            assertTrue(buffer.isBlocked().isDone());
+            assertThatThrownBy(buffer::pollPage).isEqualTo(error);
+        }
+    }
+
+    @Test
+    public void testMultipleBlockedFutures()
```

Comment:
why is that important that `isBlocked` returns different futures at each call?
Is that part of contract `ExchangeClientBuffer` provides? If so please add javadoc to interface.

Reply:
Looks it is not a contract according to `TestingExchangeClientBuffer`. So why do we test that? What would happen if single instance of future was returned?

Reply:
I'm trying to verify that cancellation of a single future doesn't affect other futures. ExchangeClient is used by multiple clients from multiple threads. If one of the clients cancels a future it should not affect other clients waiting for the `ExchangeClient`. Before it was implemented by creating one future per client that is sub-optimal. When refactoring I decided to avoid creating new futures on each call and instead use the `nonCancellationPropagating`.

Reply:
Renamed the test to `testFutureCancellationDoesNotAffectOtherFutures`